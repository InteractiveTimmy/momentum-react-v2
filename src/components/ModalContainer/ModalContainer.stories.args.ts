import { commonStyles } from '../../storybook/helper.stories.argtypes';

import { MODAL_CONTAINER_CONSTANTS as CONSTANTS } from '.';
import { PLACEMENTS } from '../ModalArrow/ModalArrow.constants';
import { DEFAULTS } from './ModalContainer.constants';

const modalContainerArgTypes = {
  showArrow: {
    description: `Whether the arrow should be shown
    <br/><br/>
    NOTE: showing the arrow and using the placement prop will not position it correctly - positioning of the arrow
    correctly will only be through the Popover component. Whenever there's the use case
    to show ModalContainer with a arrow, pls use the Popover component instead.
    <br/><br/>default \`true\``,
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: DEFAULTS.SHOW_ARROW,
      },
    },
  },
  placement: {
    description: `Placement of the Modal relative to the trigger component. 
      <br/><br/>
    NOTE: showing the arrow and using the placement prop will not position it correctly - positioning of the arrow
    correctly will only be through the Popover component. Whenever there's the use case
    to show ModalContainer with a arrow, pls use the Popover component instead.`,
    control: { type: 'select' },
    options: [undefined, ...Object.values(PLACEMENTS as Record<string, string>)],
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: DEFAULTS.PLACEMENT,
      },
    },
  },
  children: {
    description: 'Provides the child nodes for this component.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'ReactNode',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  color: {
    description: 'Provides the color of this component.',
    control: { type: 'select' },
    options: [undefined, ...Object.values(CONSTANTS.COLORS)],
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'primary',
      },
    },
  },
  elevation: {
    description:
      'Provides the elevation for this component. This only applies to the box-shadow depth.',
    control: { type: 'select' },
    options: [undefined, ...Object.values(CONSTANTS.ELEVATIONS)],
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: '0',
      },
    },
  },
  isPadded: {
    description: 'Identifies if this component should be rendered with padding.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: 'false',
      },
    },
  },
  round: {
    description:
      'Identifies if this component should be rendered with round corners and to what degree.',
    control: { type: 'select' },
    options: [undefined, ...Object.values(CONSTANTS.ROUNDS)],
    table: {
      type: {
        summary: 'number',
      },
      defaultValue: {
        summary: '0',
      },
    },
  },
};

export { modalContainerArgTypes };

export default {
  ...commonStyles,
  ...modalContainerArgTypes,
};
