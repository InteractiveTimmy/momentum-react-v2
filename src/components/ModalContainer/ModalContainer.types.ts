import { CSSProperties, ReactNode } from 'react';

export type Arrow = 'bottom' | 'left' | 'right' | 'top';
export type Color = 'primary' | 'secondary' | 'tertiary' | 'quaternary';
export type Elevation = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8;
export type Round = 0 | 25 | 50 | 75 | 100 | 125 | 150;

export interface Props {
  /**
   * Where to render the ModalArrow component.
   */
  arrow?: Arrow;

  /**
   * Child components of this ModalContainer.
   */
  children?: ReactNode;

  /**
   * Custom class for overriding this component's CSS.
   */
  className?: string;

  /**
   * Color of this ModalContainer.
   */
  color?: Color;

  /**
   * The elevation of this ModalContainer.
   */
  elevation?: Elevation;

  /**
   * Custom id for overriding this component's CSS.
   */
  id?: string;

  /**
   * If this ModalContainer is padded.
   */
  isPadded?: boolean;

  /**
   * If this ModalContainer is round and to what degree.
   *
   * Can be 0 | 25 | 50 | 75 | 100 | 125 | 150
   */
  round?: Round;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;
}
