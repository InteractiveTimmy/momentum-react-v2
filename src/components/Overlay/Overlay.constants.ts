import { Color } from './Overlay.types';

const CLASS_PREFIX = 'md-overlay';

const COLORS: Record<string, Color> = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
};

const DEFAULTS = {
  COLOR: COLORS.PRIMARY,
  FULLSCREEN: false,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, COLORS, DEFAULTS, STYLE };
