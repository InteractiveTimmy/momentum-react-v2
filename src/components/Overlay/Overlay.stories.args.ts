import {
  commonAriaDialog,
  commonAriaFocusScope,
  commonAriaOverlay,
  commonStyles,
  extendArgTypes,
} from '../../storybook/helper.stories.argtypes';
import { modalContainerArgTypes } from '../ModalContainer/ModalContainer.stories.args';

export default {
  ...commonAriaDialog,
  ...commonAriaFocusScope,
  ...extendArgTypes('ModalContainer', modalContainerArgTypes),
  ...commonAriaOverlay,
  ...commonStyles,
};
