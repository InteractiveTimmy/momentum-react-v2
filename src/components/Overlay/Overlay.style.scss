.md-overlay-wrapper {
  align-items: center;
  justify-content: center;
  display: flex;
  left: 0;
  top: 0;
  z-index: 1000;

  &[data-fullscreen='false'] {
    position: absolute;
    height: 100%;
    width: 100%;
  }

  &[data-fullscreen='true'] {
    position: fixed;
    height: 100vh;
    width: 100vw;
  }

  &[data-color='primary'] {
    background-color: var(--overlayFade-primary-background-normal);
  }

  &[data-color='secondary'] {
    background-color: var(--overlayFade-secondary-background-normal);
  }
}
