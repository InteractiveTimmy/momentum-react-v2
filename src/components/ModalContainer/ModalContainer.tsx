import React, { forwardRef, RefObject } from 'react';
import classnames from 'classnames';

import ModalArrow from '../ModalArrow';

import { ARROW_ID, DEFAULTS, STYLE } from './ModalContainer.constants';
import type { Props } from './ModalContainer.types';
import './ModalContainer.style.scss';
import { getArrowOrientation } from './ModalContainer.utils';

/**
 * The `<ModalContainer/>` component, which is used by the Popover for example to wrap
 * the content of a Popover properly.
 *
 * The `showArrow` and `placement` prop of this `<ModalContainer/>` together are only necessary
 * for the Popover component, where the Arrow and the positioning are highly important.
 * Using the `<ModalContainer/>` component standalone with the `placement` and `showArrow` props
 * will just paint, but not position the arrow correctly (done through Popover component).
 *
 * Nevertheless this component could be used on its own as well.
 */
const ModalContainer = (props: Props, ref: RefObject<HTMLDivElement>) => {
  const {
    showArrow = DEFAULTS.SHOW_ARROW,
    placement,
    children,
    elevation = DEFAULTS.ELEVATION,
    isPadded = DEFAULTS.IS_PADDED,
    round = DEFAULTS.ROUND,
    color = DEFAULTS.COLOR,
    id,
    className,
    style,
    ...otherProps
  } = props;

  const arrowOrientation = getArrowOrientation(placement);

  return (
    <div
      ref={ref}
      className={classnames(className, STYLE.wrapper)}
      id={id}
      style={style}
      data-placement={placement}
      data-arrow-orientation={arrowOrientation}
      data-color={color}
      data-elevation={elevation}
      data-padded={isPadded}
      data-round={round}
      {...otherProps}
    >
      {children}
      {
        /*arrow has to be wrapped in HTML element to allow Popover to style it*/
        showArrow && (
          <div id={ARROW_ID} data-popper-arrow className={classnames(STYLE.arrowWrapper)}>
            <ModalArrow placement={placement} color={color} />
          </div>
        )
      }
    </div>
  );
};

const _ModalContainer = forwardRef(ModalContainer);
_ModalContainer.displayName = 'ModalContainer';

export default _ModalContainer;
