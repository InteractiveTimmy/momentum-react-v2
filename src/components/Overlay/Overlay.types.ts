import { ReactNode } from 'react';
import { AriaDialogProps } from '@react-types/dialog';
import { FocusScopeProps } from '@react-aria/focus';
import { OverlayProps } from '@react-types/overlays';

import { ModalContainerProps } from '../ModalContainer';

export interface Props extends AriaDialogProps, FocusScopeProps, ModalContainerProps, OverlayProps {
  /**
   * Child components of this component. This is now a required prop.
   */
  children: ReactNode;
}
