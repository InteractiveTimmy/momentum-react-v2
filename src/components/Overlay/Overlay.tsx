import React, { forwardRef, RefObject, useRef } from 'react';
import { useOverlay } from '@react-aria/overlays';
import { useDialog } from '@react-aria/dialog';
import { mergeProps } from '@react-aria/utils';
import { FocusScope } from '@react-aria/focus';

import ModalContainer from '../ModalContainer';

import { DEFAULTS } from './Overlay.constants';
import { Props } from './Overlay.types';

/**
 * The Overlay component. This component extends `<ModalContainer />` and amends overlay functionality via React Aria.
 */
const Overlay = forwardRef((props: Props, providedRef: RefObject<HTMLDivElement>) => {
  const internalRef = useRef();

  const ref = providedRef || internalRef;

  const { overlayProps } = useOverlay(props, ref);
  const { dialogProps } = useDialog(props, ref);

  const mutatedProps = { ...props };

  const {
    autoFocus = DEFAULTS.autoFocus,
    contain = DEFAULTS.contain,
    restoreFocus = DEFAULTS.restoreFocus,
  } = mutatedProps;

  delete mutatedProps.autoFocus;
  delete mutatedProps.contain;
  delete mutatedProps.restoreFocus;

  /* eslint-disable jsx-a11y/no-autofocus */
  return (
    <FocusScope autoFocus={autoFocus} contain={contain} restoreFocus={restoreFocus}>
      <ModalContainer {...mergeProps(overlayProps, dialogProps, mutatedProps)} ref={ref} />
    </FocusScope>
  );
  /* eslint-enable jsx-a11y/no-autofocus */
});

Overlay.displayName = 'Overlay';

export default Overlay;
