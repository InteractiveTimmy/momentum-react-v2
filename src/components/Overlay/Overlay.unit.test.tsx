import React from 'react';
import { mount } from 'enzyme';
import { FocusScope } from '@react-aria/focus';

import ModalContainer from '../ModalContainer';

import Overlay, { OVERLAY_CONSTANTS as CONSTANTS } from './';

describe('<Overlay />', () => {
  const commonChildren = 'Children';

  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(<Overlay>{commonChildren}</Overlay>);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should extend <ModalContainer />', () => {
      expect.assertions(1);

      const component = mount(<Overlay>{commonChildren}</Overlay>).find(Overlay);
      const target = component.find(ModalContainer);

      expect(component.props().children).toBe(target.props().children);
    });

    it('should contain a focus scope', () => {
      expect.assertions(1);

      const component = mount(<Overlay>{commonChildren}</Overlay>).find(Overlay);

      const target = component.find(FocusScope);

      expect(target.exists()).toBe(true);
    });

    it('should pass focus scope props', () => {
      expect.assertions(1);

      const component = mount(<Overlay>{commonChildren}</Overlay>).find(Overlay);

      const target = component.find(FocusScope);

      expect(target.props()).toMatchObject(CONSTANTS.DEFAULTS);
    });
  });
});
