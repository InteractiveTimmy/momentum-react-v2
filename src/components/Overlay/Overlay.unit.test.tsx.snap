// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overlay /> snapshot should match snapshot 1`] = `
<Overlay>
  <FocusScope
    autoFocus={false}
    contain={false}
    restoreFocus={false}
  >
    <span
      data-focus-scope-start={true}
      hidden={true}
    />
    <ModalContainer
      aria-labelledby={null}
      onKeyDown={[Function]}
      role="dialog"
      tabIndex={-1}
    >
      <div
        aria-labelledby={null}
        className="md-modal-container-wrapper"
        onKeyDown={[Function]}
        role="dialog"
        tabIndex={-1}
      >
        <div
          className="md-modal-container-content"
          data-color="primary"
          data-elevation={0}
          data-horizontal-arrow={false}
          data-padded={false}
          data-round={0}
          data-vertical-arrow={false}
        >
          Children
        </div>
      </div>
    </ModalContainer>
    <span
      data-focus-scope-end={true}
      hidden={true}
    />
  </FocusScope>
</Overlay>
`;
